### PR TITLE
state-root comparison

### PR DESCRIPTION
**Pull request by cassc**
Export states to allow goevmlab to compare the stateroot of transactions with other EVM implementations. 

Test file used: [sstore-input.json](https://github.com/sbip-sg/CuEVM/files/14900878/sstore-input.json)

- [x] print list of all accounts as json list to stderr 
- [x] test we get same stateroot hash with `geth`
  - [x] check why accounts in state from cuevm are different from `geth` 
  - [x] hash the stateroot in govemlab
- [x] update test script to allow comparing stateroot hash
- [x] post transaction state should contain the `touch_state` 
- [x] refunded gas computation bug in `sstore` 
- [x] value in contract creation transtion not deducted from sender
- [x] double free issue caused by mixing CPP destructors and manual pointers. ~Unresolved in this PR. This affects the test only.~ Some other mem leaks still reported by valgrind in `update_with_child_state`

Possible sources of stateRoot test failures if the traces matches:

- Check all accounts in the previous states have been considered
- Refunded gas computation 
- account `nonce` 
- account `balance` error 
- if all the above matches, yet the stateRoot still mismatches, then it could be a computation bug in cuevm driver in the goevmlab

Run tests against an input folder with test files:
```
python ~/projects/sbip-sg/CuEVM/scripts/run-ethtest-by-fork.py -t /tmp/out --runtest-bin runtest --geth geth --cuevm cuevm-path  -i input-folder-path
```

*There are more stateRoot test failures, I am not attempting to solve them in this PR.*

Run [fuzz-test](https://github.com/cassc/goevmlab) (recommended as it's more efficient in finding differences):

```
generic-fuzzer --geth geth --cuevm ~/projects/sbip-sg/CuEVM/out/cpu_debug_interpreter --fork Shanghai
```

To debug any trace mismatch, you can compare the `jsonl` files generated by goevmlab by `generic-fuzzer` or `runtest`.

To inpect any stateRoot mismatch when the traces match, compare the statedump from geth, 
```
geth --json --noreturndata --dump statetest [input-json-file]
# states are printed at the end of the stdout
```

with cuevm states from, 

```
cuevm --input [input-json-file] --output /dev/null
# states are printed at the end of the stdout
```

Thes states have to inspected manually for the differences.

### CODE REVIEW

**Comment by cassc** [comment]
Difference found in the stateRoot comparison:

Test input: [input.json](https://github.com/sbip-sg/CuEVM/files/14681577/cuevm-input.json)

Output jsons copied from the outputs of:

```bash
geth --json --noreturndata --dump --nomemory statetest input.json
cuevm  --output /dev/null --input   /tmp/out/./sstore_combinations_initial11-0-0-0/sstore_combinations_initial11.json
```



```
================================================================================
GETH
address: 0x2000000000000000000000000000000000000000, storageRoot: 0xf38f9f63c760d088d7dd04f743619b6291f63beebd8bdf530628f90e9cfa52d7
address: 0x3000000000000000000000000000000000000000, storageRoot: 0xb160f402dd1b93be873fa7f5f952b56ccdaa3cc2f02c9786f02978afef4b461a
address: 0x6295eE1B4F6dD65047762F924Ecd367c17eaBf8f, storageRoot: 0xbf3e65b043d6d870557f8556faf48fd77aecfd8558e3b8087ee399b3a8d62694
address: 0xB000000000000000000000000000000000000000, storageRoot: 0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421
address: 0xB100000000000000000000000000000000000000, storageRoot: 0x275d689ae4899d1f5487644614def6d00cd13efdd3938b5ccfa450aa49bcddf5
address: 0xB200000000000000000000000000000000000000, storageRoot: 0xb160f402dd1b93be873fa7f5f952b56ccdaa3cc2f02c9786f02978afef4b461a
address: 0xa94f5374Fce5edBC8E2a8697C15331677e6EbF0B, storageRoot: 0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421

================================================================================
CUEVM
address: 0x1000000000000000000000000000000000000000, storageRoot: 0x8bc7a2dd59c478c128505931d4ea85d58531694f06e72338b9e8f0d32151b722
address: 0x3000000000000000000000000000000000000000, storageRoot: 0x936c8fc346f3a026b98409b9c9c947323878bfb6b2a110812c15966191bd84de
address: 0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b, storageRoot: 0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421
address: 0xb100000000000000000000000000000000000000, storageRoot: 0xe12c820d4a34b8ceb99badf9668af8e2b86e9b6e388989319bfcb280ad180022
address: 0xb200000000000000000000000000000000000000, storageRoot: 0x936c8fc346f3a026b98409b9c9c947323878bfb6b2a110812c15966191bd84de
CUEVM state_root 0x8c77d9d74d6ba95d9bf1d128e4166f07da1ace4bdd80699440a17c48051620c8
```

**Comment by cassc** [comment]
`0x6295eE1B4F6dD65047762F924Ecd367c17eaBf8f` is created by contract deployment transction, it's missing from CUEVM, perhaps we are missing some world state update? @sdcioc

**Comment by sdcioc** [comment]
> `0x6295eE1B4F6dD65047762F924Ecd367c17eaBf8f` is created by contract deployment transction, it's missing from CUEVM, perhaps we are missing some world state update? @sdcioc

We consider the "pre" part of the transaction to be the part of the world state previous to the transaction. It looks like they consider a null world state, and before the transaction, they deploy the "pre" contracts. Also, the final touch state contains only the modification to the world state, not the final world state of the respective accounts. to compute the final storage, you need to take the account from the world state and then make the modifications from the touch state. `0x6295eE1B4F6dD65047762F924Ecd367c17eaBf8f` I have to look and test. It is strange that the coinbase address is not included in the modifications.

File: src/state.cuh [hunk 1]
```diff
@@ -1785,6 +1873,18 @@ public:
...
 
+    __host__ __device__ __forceinline__ touch_state_t(
+        touch_state_data_t *content,
+        accessed_state_t *access_state,
+        arith_t &arith
+    ) : _arith(arith),
+        _content(content),
+        _accessed_state(access_state),
+        _parent_state(nullptr),
+        nodestruct(true)
```
Comment (cassc) [inline src/state.cuh:1884]: `nodestruct` is used to attempt to solve the double free bug, not working

File: src/evm.cuh [hunk 2]
```diff
@@ -3406,8 +3412,119 @@ public:
...
+
+
+        #endif
+        if (cpu_world_state != nullptr){
+            delete cpu_world_state;
+            cpu_world_state = NULL;
+        }
+    }
+
+    __host__ static void print_touched_trie_accounts(arith_t &arith, touch_state_data_t *evm_instances)    {
```
Comment (cassc) [inline src/evm.cuh:3455]: these test or debug only functions should be placed outside the current file which contains core EVM logic to reduce cluttering. Any suggestions?

**Pull request closed without merging**